File: harvester/parsers/splitting/registry.py
"""Registry for element hierarchy specifications."""

from __future__ import annotations

from harvester.parsers.splitting.protocols import ElementSpec


class HierarchyRegistry:
    """Registry for element hierarchy specifications.

    Maps XML tag names to their ElementSpec definitions, providing
    a declarative way to describe the document structure.
    """

    def __init__(self) -> None:
        """Initialize an empty registry."""
        self._specs: dict[str, ElementSpec] = {}

    def register(self, spec: ElementSpec) -> None:
        """Register an element specification.

        Args:
            spec: The element specification to register
        """
        self._specs[spec.tag] = spec

    def get_spec(self, tag: str) -> ElementSpec | None:
        """Get specification for a tag.

        Args:
            tag: The tag name (without namespace)

        Returns:
            The element specification, or None if not registered
        """
        return self._specs.get(tag)

    def is_structural(self, tag: str) -> bool:
        """Check if tag is a registered structural element.

        Args:
            tag: The tag name

        Returns:
            True if the tag has a registered specification
        """
        return tag in self._specs

    def registered_tags(self) -> set[str]:
        """Return set of all registered tag names."""
        return set(self._specs.keys())

File: packages/harvester/src/splitting/registry.rs
//! Hierarchy registry for article splitting.

use std::collections::HashMap;

use super::types::ElementSpec;

/// Registry of element specifications for the hierarchy.
pub struct HierarchyRegistry {
    specs: HashMap<String, ElementSpec>,
}

impl HierarchyRegistry {
    /// Create a new empty registry.
    #[must_use]
    pub fn new() -> Self {
        Self {
            specs: HashMap::new(),
        }
    }

    /// Register an element specification.
    pub fn register(&mut self, spec: ElementSpec) {
        self.specs.insert(spec.tag.clone(), spec);
    }

    /// Get the specification for a tag.
    #[must_use]
    pub fn get_spec(&self, tag: &str) -> Option<&ElementSpec> {
        self.specs.get(tag)
    }

    /// Check if a tag is a structural element in the hierarchy.
    #[must_use]
    pub fn is_structural(&self, tag: &str) -> bool {
        self.specs.contains_key(tag)
    }
}

impl Default for HierarchyRegistry {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_registry_register_and_get() {
        let mut registry = HierarchyRegistry::new();

        let spec = ElementSpec::new("artikel").with_split_point(true);
        registry.register(spec);

        let retrieved = registry.get_spec("artikel");
        assert!(retrieved.is_some());
        assert!(retrieved.unwrap().is_split_point);
    }

    #[test]
    fn test_registry_is_structural() {
        let mut registry = HierarchyRegistry::new();
        registry.register(ElementSpec::new("lid"));

        assert!(registry.is_structural("lid"));
        assert!(!registry.is_structural("al"));
    }
}
